File: Workspace/Feedback Control for Computer Systems/src/main/java/nl/tudelft/rvh/chapter1/BufferClosed.java
package nl.tudelft.rvh.chapter1;

import java.util.concurrent.TimeUnit;

import nl.tudelft.rvh.ChartTab;
import nl.tudelft.rvh.Tuple;
import rx.Observable;
import rx.functions.Func1;
import rx.functions.Func2;
import rx.subjects.PublishSubject;

public class BufferClosed extends ChartTab {

	public BufferClosed() {
		super("Chapter 1 - closed", "Closed simulation of random buffer simulation",
				"time", "queue size");
	}

	@Override
	public String seriesName() {
		return "closed simulation";
	}

	@Override
	public Observable<Tuple<Number, Number>> runSimulation() {
		Observable<Long> time = Observable.interval(1L, TimeUnit.MILLISECONDS)
				.take(5000);
		Buffer b = new Buffer(10, 10);
		Func1<Long, Integer> target = t -> t < 100 ? 0
				: t < 300 ? 50
						: 10;
		Func2<Integer, Integer, Double> control = (e, c) -> 1.25 * e + 0.01 * c;

		Observable<Integer> feedbackLoop = Observable.create(subscriber -> {
			PublishSubject<Integer> queueLength = PublishSubject.create();

			Observable<Integer> error = time.map(target).zipWith(queueLength, Math::subtractExact);

			error.scan(new Tuple<>(0, 0), (c, e) -> new Tuple<>(e, e + c.getY()))
					.map(t -> control.call(t.getX(), t.getY()))
					.map(b::work)
					.subscribe(queueLength::onNext);

			queueLength.subscribe(subscriber);
			queueLength.onNext(0);
		});
		return time.zipWith(feedbackLoop.onBackpressureBuffer().skip(1),
				Tuple<Number, Number>::new);
	}
}

File: Workspace/Feedback Control for Computer Systems/src/main/scala/nl/tudelft/rvh/chapter1/BufferClosed.scala
package nl.tudelft.rvh.chapter1

import scala.concurrent.duration.DurationInt

import nl.tudelft.rvh.ChartTab
import rx.lang.scala.Observable
import rx.lang.scala.schedulers.ComputationScheduler
import rx.lang.scala.subjects.BehaviorSubject

class BufferClosed extends ChartTab("Chapter 1 - closed", "Closed simulation of random buffer simulation", "time", "queue size") {

	def seriesName = "closed simulation"

	override def time = Observable interval (1 millisecond, ComputationScheduler()) take 5000
	
	def setpoint(t: Long) = if (t < 100) 0 else if (t < 300) 50 else 10
	
	def simulation = {
		val buffer = new Buffer(10, 10)
		def control(e: Double, c: Double) = 1.25 * e + 0.01 * c
		
		Observable[Int](subscriber => {
			val queueLength = BehaviorSubject(0)
			queueLength subscribe subscriber
			
			time.map(setpoint)
				.zipWith(queueLength)(_ - _)
				.scan((0.0, 0.0))((c, e) => (e, e + c._2))
				.map((control(_, _)) tupled)
				.map(buffer work)
				.subscribe(queueLength)
		}).onBackpressureBuffer drop 1
	}
}